### PR TITLE
[Tvnow] bugfix & download whole serie

### PR DESCRIPTION
**Pull request by AndroKev**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

with this pull request you can download a whole serie

### CODE REVIEW

File: youtube_dl/extractor/tvnow.py [hunk 1]
```diff
@@ -178,7 +178,47 @@ def _real_extract(self, url):
                 if not seo_url:
                     continue
                 entries.append(self.url_result(
-                    base_url + seo_url + '/player', 'TVNow', info.get('id')))
+                    base_url + seo_url + '/player', 'TVNow', compat_str(info.get('id'))))
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Won't work on `None`.

File: youtube_dl/extractor/tvnow.py [hunk 2]
```diff
@@ -178,7 +178,47 @@ def _real_extract(self, url):
...
 
         return self.playlist_result(
             entries, compat_str(season.get('id') or season_id), title)
+
+
+class TVNowListChannelIE(TVNowBaseIE):
+    _VALID_URL = r'(?P<base_url>https?://(?:www\.)?tvnow\.(?:de|at|ch)/(?:rtl(?:2|plus)?|nitro|superrtl|ntv|vox)/(?P<show_id>[^/]+))$'
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: No trailing `$`. Override `suitable`.

File: youtube_dl/extractor/tvnow.py [hunk 3]
```diff
@@ -178,7 +178,47 @@ def _real_extract(self, url):
...
+        entries = []
+        for season_info in list_info['formatTabs']['items']:
+            season_url = season_info.get('seoheadline')
+            if not season_url:
+                continue
+            entries.append(self.url_result(
+                base_url + "/list/" + season_url, 'TVNowList', compat_str(season_info.get('id')), season_info.get('headline')))
+
+        return self.playlist_result(
+            entries, compat_str(list_info['id']), compat_str(list_info['title'] or show_id))
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Playlist metadata must not be fatal.

File: youtube_dl/extractor/tvnow.py [hunk 4]
```diff
@@ -178,7 +178,47 @@ def _real_extract(self, url):
...
+
+        fields = []
+        fields.extend(self._SHOW_FIELDS)
+        fields.extend('formatTabs.%s' % field for field in self._SEASON_FIELDS)
+
+        list_info = self._call_api(
+            'formats/seo', show_id, query={
+                'fields': ','.join(fields),
+                'name': show_id + '.php'
+            })
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Code duplication.

File: youtube_dl/extractor/tvnow.py [hunk 5]
```diff
@@ -178,7 +178,47 @@ def _real_extract(self, url):
...
+        },
+        'playlist_mincount': 1,
+    }]
+
+    def _real_extract(self, url):
+        base_url, show_id = re.match(self._VALID_URL, url).groups()
+
+        fields = []
+        fields.extend(self._SHOW_FIELDS)
+        fields.extend('formatTabs.%s' % field for field in self._SEASON_FIELDS)
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Code duplication.

**Comment by AndroKev** [comment]
hope now it's like you want it

File: youtube_dl/extractor/tvnow.py [hunk 6]
```diff
@@ -178,7 +186,40 @@ def _real_extract(self, url):
                 if not seo_url:
                     continue
                 entries.append(self.url_result(
-                    base_url + seo_url + '/player', 'TVNow', info.get('id')))
+                    base_url + seo_url + '/player', 'TVNow', str(info.get('id'))))
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Nothing changed.

File: youtube_dl/extractor/tvnow.py [hunk 7]
```diff
@@ -178,7 +186,40 @@ def _real_extract(self, url):
...
 
         return self.playlist_result(
             entries, compat_str(season.get('id') or season_id), title)
+
+
+class TVNowListChannelIE(TVNowBaseIE):
+    _VALID_URL = r'(?P<base_url>https?://(?:www\.)?tvnow\.(?:de|at|ch)/(?:rtl(?:2|plus)?|nitro|superrtl|ntv|vox)/(?P<show_id>[^/]+)$)'
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: I've already pointed out: no trailing `$`. Override `suitable` instead.

File: youtube_dl/extractor/tvnow.py [hunk 8]
```diff
@@ -152,18 +170,8 @@ class TVNowListIE(TVNowBaseIE):
...
-        fields.extend(
-            'formatTabs.formatTabPages.container.movies.%s' % field
-            for field in self._VIDEO_FIELDS)
-
-        list_info = self._call_api(
-            'formats/seo', season_id, query={
-                'fields': ','.join(fields),
-                'name': show_id + '.php'
-            })
+        fields = self._extend_query(self._SHOW_FIELDS, self._SEASON_FIELDS, self._VIDEO_FIELDS)
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Code duplication 173, 213. There is no sense to extract fields explicitly.

File: youtube_dl/extractor/tvnow.py [hunk 9]
```diff
@@ -26,6 +26,24 @@ def _call_api(self, path, video_id, query):
...
+        fields.extend(show)
+        fields.extend('formatTabs.%s' % field for field in season)
+        if video:
+            fields.extend(
+                'formatTabs.formatTabPages.container.movies.%s' % field
+                for field in video)
+
+        return fields
+
+    def _tvnow_list_info(self, list_id, show_id, fields):
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: This is not used in single video extractor thus should not be here.

File: youtube_dl/extractor/tvnow.py [hunk 10]
```diff
@@ -178,7 +186,40 @@ def _real_extract(self, url):
...
+        entries = []
+        for season_info in list_info['formatTabs']['items']:
+            season_url = season_info.get('seoheadline')
+            if not season_url:
+                continue
+            entries.append(self.url_result(
+                base_url + "/list/" + season_url, 'TVNowList', compat_str(season_info.get('id')), season_info.get('headline')))
+
+        return self.playlist_result(
+            entries, list_info['id'], list_info['title'])
```
Comment (dstftw) [inline youtube_dl/extractor/tvnow.py]: Playlist id and title should not be fatal.

**Pull request closed without merging**